pkg/scope/health/health.go

```diff
@@ -0,0 +1,148 @@
+	errResourceNotFound      = "could not retrieve resource %q %q %q"
+	errDeploymentUnavailable = "no ready instance found in %q %q %q"
+
+	defaultTimeout = 10 * time.Second
+)
+
+// UpdateHealthStatus updates the status of the healthscope based on workload resources.
+func UpdateHealthStatus(ctx context.Context, client client.Client, healthScope *v1alpha2.HealthScope) error {
+	timeout := defaultTimeout
+	if healthScope.Spec.ProbeTimeout != nil {
+		timeout = time.Duration(*healthScope.Spec.ProbeTimeout) * time.Second
+	}
+
+	resourceRefs := []runtimev1alpha1.TypedReference{}
+	for _, workloadRef := range healthScope.Status.WorkloadReferences {
+		// Get workload object.
+		workloadObject := unstructured.Unstructured{}
+		workloadObject.SetAPIVersion(workloadRef.APIVersion)
+		workloadObject.SetKind(workloadRef.Kind)
+		workloadObjectRef := types.NamespacedName{Namespace: healthScope.GetNamespace(), Name: workloadRef.Name}
+		if err := client.Get(ctx, workloadObjectRef, &workloadObject); err != nil {
+			return errors.Wrapf(err, errNoWorkload, workloadRef.Name)
+		}
+
+		// TODO(asouza): Need to find a way to make sure that this is compatible with the scope structure.
+		if value, err := fieldpath.Pave(workloadObject.UnstructuredContent()).GetValue("status.resources"); err == nil {
+			refs := value.([]interface{})
+			for _, item := range refs {
+				ref := item.(map[string]interface{})
+				resourceRef := runtimev1alpha1.TypedReference{
+					APIVersion: fmt.Sprintf("%v", ref["apiVersion"]),
+					Kind:       fmt.Sprintf("%v", ref["kind"]),
+					Name:       fmt.Sprintf("%v", ref["name"]),
+				}
+
+				resourceRefs = append(resourceRefs, resourceRef)
+			}
+		} else {
+			return errors.Wrapf(err, errNoWorkloadResources, workloadRef.Name)
+		}
+	}
+
+	statusc := resourcesHealthStatus(ctx, client, healthScope.Namespace, resourceRefs, timeout)
+	status := true
+	for r := range statusc {
+		status = status && r
+	}
+
+	health := "unhealthy"
+	if status {
+		health = "healthy"
+	}
+
+	healthScope.Status.Health = health
+	return nil
+}
+
+func resourcesHealthStatus(ctx context.Context, client client.Client, namespace string, refs []runtimev1alpha1.TypedReference, timeout time.Duration) <-chan bool {
+	status := make(chan bool)
+	var wg sync.WaitGroup
+	wg.Add(len(refs))
+	for _, ref := range refs {
+		go func(resourceRef runtimev1alpha1.TypedReference) {
+			defer wg.Done()
+			err := resourceHealthStatus(ctx, client, namespace, resourceRef, timeout)
+			status <- (err == nil)
+			if err != nil {
+				fmt.Printf("Unhealthy resource %q: %v\n", resourceRef.Name, err)
+			}
+		}(ref)
+	}
+	go func() {
+		wg.Wait()
+		close(status)
+	}()
+
+	return status
+}
+
+func resourceHealthStatus(ctx context.Context, client client.Client, namespace string, ref runtimev1alpha1.TypedReference, _ time.Duration) error {
+	// TODO(artursouza): figure out how to apply timeout.
+	if (ref.APIVersion == "apps/v1") && (ref.Kind == "Deployment") {
```

Comment:
you can get those programmatically instead of hard code

Reply:
Do you have an example?

Reply:
```suggestion
	if ref.GroupVersionKind() == appsv1.SchemeGroupVersion.WithKind("Deployment") {
```

Will more common

Reply:
Done.